### PR TITLE
Add support for encrypting gossip traffic

### PR DESCRIPTION
**Pull request by sanimej**
Signed-off-by: Santhosh Manohar <santhosh@docker.com>

### CODE REVIEW

File: networkdb/cluster.go [hunk 1]
```diff
@@ -33,6 +34,38 @@ func (l *logWriter) Write(p []byte) (int, error) {
...
+	nDB.keyring.AddKey(key)
+}
+
+func (nDB *NetworkDB) SetPrimaryKey(key []byte) {
+	for _, dbKey := range nDB.config.Keys {
+		if bytes.Equal(key, dbKey) {
+			if nDB.memberlist != nil {
+				nDB.keyring.UseKey(dbKey)
+			}
+		}
```
Comment (aboch) [inline networkdb/cluster.go]: Should you stop iterating, once you find the target key ?

File: networkdb/cluster.go [hunk 2]
```diff
@@ -33,6 +34,38 @@ func (l *logWriter) Write(p []byte) (int, error) {
...
+
+func (nDB *NetworkDB) RemoveKey(key []byte) {
+	for i, dbKey := range nDB.config.Keys {
+		if bytes.Equal(key, dbKey) {
+			nDB.config.Keys = append(nDB.config.Keys[:i], nDB.config.Keys[i+1:]...)
+
+			if nDB.memberlist != nil {
+				nDB.keyring.RemoveKey(dbKey)
+			}
+		}
```
Comment (aboch) [inline networkdb/cluster.go]: Here too

File: networkdb/cluster.go (hunk 1 above)
Comment (sanimej) [inline networkdb/cluster.go]: Yes, its ok to stop iterating after finding the added or removed key.

I did the same in agent.go/agentHandleKeys(). It also removes the slice entry deletion while iterating over it. PTAL.

**Comment by aboch** [comment]
LGTM

**Comment by mrjana** [comment]
LGTM

**Pull request merged**